keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,389 @@
+   // Generally, labels here should point to a specific pod on a node.
+   Selector LabelSelector `json:"selector,omitempty" protobuf:"bytes,2,opt,name=selector"`
+   // The namespace query for the pods representing this readiness gate.
+   Namespace string `json:"namespace,omitempty" protobuf:"bytes,3,opt,name=namespace"`
+   // Indicates if this readiness gate should block the overall readiness of a node. Default: true
+   // +optional
+   BlocksReadiness bool `json:"blocksReadiness,omitempty" protobuf:"varint,4,opt,name=blocksReadiness"`
+}
+```
+
+For each `NodeReadinessGate`, kubelet will fetch the state of all pods that match its label selector and take the readiness state of those pods into account.
+If all pods represented by every `NodeReadinessGate` is ready (based on container state and readiness probes) then the node is also considered ready.
+If no pods are found with the label selector then the node is consider to be "Not Ready".
+
+The individual pods representing readiness gates are required to detect kubelet crash (or restart) quickly and reflect that in their own readiness states via probes.
+The kubelet places a special meta readiness taint on a node object if 1 or more of those pods are not ready. If the Node Lifecycle Controller observes that all the pods selected
+by `ReadinessGates` are ready, it will remove that taint immediately.
+
+If `BlocksReadiness` is set to `false`, the overall readiness of a node is not blocked by the pods under this specific readiness gate. This is useful for non-critical pods
+that should not block the readiness of a node but should be reported in the node status.
+
+### Readiness Gates from External Sources
+
+The `ReadinessGates` field can be used by external components to dictate node readiness since they can apply an arbitrary readiness gate with a reserved label selector
+that should not be used by any pods on the cluster. A common usecase here is an external cloud provider that wants to ensure no pods are scheduled on a node until
+networking for a given node is properly set up by the cloud provider. In this case, the cloud provider would apply any custom readiness gates to a node during registration
+and a separate controller would remove the readiness gate when it sees that node's networking has been properly configured. No pods matching the label selector of that
+readiness gate was required.
+
+### Reporting Node Readiness via Node Status
+
+The readiness of a node will be reported in the node's status subresource:
+
+```go
+type NodeStatus struct {
+    ...
+    Readiness []NodeReadiness `json:"readiness,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,12,rep,name=readiness"`
+}
+
+type NodeReadiness struct {
+    // The name of the node readiness gate
+    // +optional
+    Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+    // Whether a node is ready based on it's corresponding readiness gates
```

Comment:
"its". But also, "gates" plural makes it sound like this is the aggregated ready state, but it's not. "`Whether the node is ready based on this readiness gate`".
And neither field should be `+optional` right?

Reply:
Good catch, updated.